js/A step by step guide to functions in JavaScript

```diff
@@ -0,0 +1,213 @@
+If you declared a parameter, but did not pass an argument to it, your parameter would be undefined.
+In JavaScript, function parameters default to *undefined*.
+
+
+### 📌 Function arguments
+The argument is the actual value that gets passed into the function.
+
+![functionblog1.png](https://cdn.hashnode.com/res/hashnode/image/upload/v1587151140206/6UMKRMJdW.png)
+
+*You define a function with parameters, you call a function with arguments.* In function sum() a and b are parameters whereas 5 and 3 are arguments.
+
+### 📌 The return Statement
+A JavaScript function can have an optional return statement. This is required if you want to return a value from a function. This statement should be the last statement in a function. The return statement stops the execution of a function and returns a value from that function.
+A function always returns a value, If the return value is not specified then *undefined* is returned.
+
+## The function expression 
+A function expression looks similar to function declarations, except that the function is assigned to a variable name. 
+```javascript
+var sum=function(a,b){
+return a+b;
+};
+
+sum(5,3);
+//output : 8
+```
+Functions stored in variables do not need function names. They are always invoked (called) using the variable name. The function above ends with a semicolon because it is a part of an executable statement.
+
+## Hoisting 
+Hoisting was thought up as a general way of thinking about how execution contexts work in JavaScript. Hoisting means variable and function declarations are moved to the top of scope before execution. It allows you to use a function before you declare it in your code.
+
+### 📌 Difference between function declaration and function expression
+> 
+Function declarations are hoisted but function expressions are not. 
+
+Functions defined in a function declaration are hoisted, which means that you can use the function although it's defined below the code using it. Hoisted functions are made available everywhere within the current scope. For example
+
+```javascript
+//function is invoked before declaration
+sum(5,3);  //output : 8
+//function declaration
+function sum(a,b){
+return a+b;
+}
+
+```
+As opposed to function declarations, function expressions are not hoisted and can therefore not be used before they're defined. 
+```javascript
+sum(5,3);  //output :  // TypeError: sum is not a function
+//function expression
+var sum =function(a,b){
+return a+b;
+}
+```
+## IIFE (Immediately Invoked Function Expressions)
+Functions that are executed as soon as they are declared, these functions are known as Immediately Invoked Function Expressions or IIFEs. 
+IIFEs follows a particular syntax as shown below.
+```javascript
+(function (){ 
+// code to be executed
+})();
+```
+Let's break it down to make more sense. We basically have a function defined inside parentheses, and then we append () to execute that function.
```

Comment:
basically is something that we often use while talking, you can skip this in writing.  * We have a function defined inside parentheses,..*